functions/__bass.py

```diff
@@ -14,127 +14,216 @@
-            cmd=' '.join(sys.argv[1:]),
-            output=new_env + alias_str
-        )
-    new_env = new_env.strip()
+        raise subprocess.CalledProcessError(p.returncode, p.args)
+
+    new_env = load_env(new_env)
+    old_bash_state, _ = load_state(old_bash_state)
+    new_bash_state, function_source = load_state(new_bash_state)
 
-    old_env = json.loads(old_env)
-    new_env = json.loads(new_env)
+    # save the bash internal state in this variable so we can immediately resurrect it next time
+    saved_bash_state = b''
+    for line in new_bash_state:
+        saved_bash_state += line + b'\n'
+    saved_bash_state += function_source
+    with open(state_file, 'wb') as f:
+        f.write(saved_bash_state)
 
     script_lines = []
 
+    # env vars
     for k, v in new_env.items():
         if ignored(k):
             continue
         v1 = old_env.get(k)
         if not v1:
-            script_lines.append(comment('adding %s=%s' % (k, v)))
+            script_lines.append(comment(b'adding %s=%s' % (k, v)))
         elif v1 != v:
-            script_lines.append(comment('updating %s=%s -> %s' % (k, v1, v)))
+            script_lines.append(comment(b'updating %s=%s -> %s' % (k, v1, v)))
             # process special variables
-            if k == 'PWD':
-                script_lines.append('cd %s' % escape(v))
+            if k == b'PWD':
+                script_lines.append(b'cd %s' % escape(v))
                 continue
         else:
             continue
-        if k == 'PATH':
-            value = ' '.join([escape(directory)
-                              for directory in v.split(':')])
+        if k == b'PATH':
+            value = b' '.join([escape(directory)
+                              for directory in v.split(b':')])
         else:
             value = escape(v)
-        script_lines.append('set -g -x %s %s' % (k, value))
+        script_lines.append(b'set -g -x %s %s' % (escape(k), value))
 
     for var in set(old_env.keys()) - set(new_env.keys()):
-        script_lines.append(comment('removing %s' % var))
-        script_lines.append('set -e %s' % var)
+        script_lines.append(comment(b'removing %s' % var))
+        script_lines.append(b'set -e %s' % escape(var))
 
-    script = '\n'.join(script_lines)
+    # aliases
+    old_aliases = parse_aliases(old_bash_state)
+    new_aliases = parse_aliases(new_bash_state)
 
-    alias_lines = []
-    for line in alias_str.splitlines():
-        _, rest = line.split(None, 1)
-        k, v = rest.split("=", 1)
-        alias_lines.append("alias " + escape_identifier(k) + "=" + v)
-    alias = '\n'.join(alias_lines)
+    for k, v in new_aliases.items():
+        v1 = old_aliases.get(k)
+        if not v1:
+            script_lines.append(comment(b'adding alias %s=%s' % (k, v)))
+        elif v1 != v:
+            script_lines.append(comment(b'updating alias %s=%s -> %s' % (k, v1, v)))
+        else:
+            continue
+        script_lines.append(b'alias %s %s' % (k, v))
```

Comment:
Aliases probably need to be run through `bass` as well, don't they? At least, I'm currently encountering an alias that contains bash code.